### PR TITLE
feat(snownet): only upgrade socket if it is actually an upgrade

### PR DESCRIPTION
**Pull request by thomaseizinger**
`str0m` may issue us a series of `NominatedSend` events that are all "equal" in the sense that they aren't different from a network topology PoV.

In some testing using iperf, I've found this to slightly reduce latency. It is still around 7 seconds with a cold-start so more improvements are needed.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
I just realised that a consequence of this PR is that we can't "downgrade" to a relayed connection if e.g. the current direct one fails.

I think we might actually need an event from str0m here where we can react to pairs getting "unnominated".

File: rust/connlib/snownet/src/node.rs [hunk 1]
```diff
@@ -1178,13 +1190,67 @@ enum PeerSocket {
...
+        (CandidateKind::ServerReflexive, CandidateKind::Host) => Less,
+        (CandidateKind::Relayed, CandidateKind::Host) => Less,
+        (CandidateKind::Relayed, CandidateKind::PeerReflexive) => Less,
+        (CandidateKind::Relayed, CandidateKind::ServerReflexive) => Less,
+        (CandidateKind::Host, CandidateKind::Host)
+        | (CandidateKind::Relayed, CandidateKind::Relayed)
+        | (CandidateKind::PeerReflexive, CandidateKind::PeerReflexive)
+        | (CandidateKind::ServerReflexive, CandidateKind::ServerReflexive)
+        | (CandidateKind::ServerReflexive, CandidateKind::PeerReflexive)
+        | (CandidateKind::PeerReflexive, CandidateKind::ServerReflexive) => Equal,
```
Comment (conectado) [inline rust/connlib/snownet/src/node.rs:1250]: Maybe we should prefer peer reflexive, as the [ICE RFC describes](https://datatracker.ietf.org/doc/html/rfc8445#appendix-B.7)

**Review by conectado — APPROVED** [review]

**Comment by thomaseizinger** [comment]
> I just realised that a consequence of this PR is that we can't "downgrade" to a relayed connection if e.g. the current direct one fails.
> 
> I think we might actually need an event from str0m here where we can react to pairs getting "unnominated".

I"ve been talking to str0m maintainers and they say this sounds like a bug that should be handled in str0m.

**Comment by thomaseizinger** [comment]
> > I just realised that a consequence of this PR is that we can't "downgrade" to a relayed connection if e.g. the current direct one fails.
> > I think we might actually need an event from str0m here where we can react to pairs getting "unnominated".
> 
> I"ve been talking to str0m maintainers and they say this sounds like a bug that should be handled in str0m.

Bug fix here: https://github.com/algesten/str0m/pull/463

**Pull request closed without merging**